File: agent/library/src/main/java/org/matilda/messages/listener/MessageListeningInstance.java
package org.matilda.messages.listener;

import org.matilda.messages.Message;
import org.matilda.messages.MessageHandlerRegistration;

import java.util.concurrent.BlockingQueue;

public class MessageListeningInstance implements AutoCloseable {
    private final BlockingQueue<Message> mMessageQueue;
    private final MessageHandlerRegistration mRegistration;

    public MessageListeningInstance(BlockingQueue<Message> messageQueue, MessageHandlerRegistration registration) {
        mMessageQueue = messageQueue;
        mRegistration = registration;
    }

    public Message waitForMessage() throws InterruptedException {
        return mMessageQueue.take();
    }

    public void stop() {
        mRegistration.unregister();
    }

    public void close() {
        stop();
    }
}

File: agent/library/src/main/java/org/matilda/messages/listener/MessageListeningInstance.kt
package org.matilda.messages.listener

import kotlinx.coroutines.channels.Channel
import org.matilda.messages.Message
import org.matilda.messages.MessageHandlerRegistration
import java.io.Closeable

class MessageListeningInstance(
    private val mMessageChannel: Channel<Message>,
    private val mRegistration: MessageHandlerRegistration
) : Closeable {
    suspend fun waitForMessage(): Message {
        return mMessageChannel.receive()
    }

    fun stop() {
        mRegistration.unregister()
    }

    override fun close() {
        stop()
    }
}
